packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js

```diff
@@ -0,0 +1,294 @@
+          });
+          jest.runAllTimers();
+        });
+
+        expect(container.innerHTML).toBe('1');
+      });
+    });
+
+    it('warns if you return a value inside act', () => {
+      expect(() => act(() => null)).toWarnDev(
+        [
+          'The callback passed to ReactTestUtils.act(...) function must return undefined, or a Promise.',
+        ],
+        {withoutStack: true},
+      );
+      expect(() => act(() => 123)).toWarnDev(
+        [
+          'The callback passed to ReactTestUtils.act(...) function must return undefined, or a Promise.',
+        ],
+        {withoutStack: true},
+      );
+    });
+
+    it('warns if you try to await an .act call', () => {
+      expect(act(() => {}).then).toWarnDev(
+        [
+          'Do not await the result of calling ReactTestUtils.act(...) with sync logic, it is not a Promise.',
+        ],
+        {withoutStack: true},
+      );
+    });
+  });
+  describe('async', () => {
+    it('does the async stuff', async () => {
+      function App() {
+        let [ctr, setCtr] = React.useState(0);
+        function doSomething() {
+          setTimeout(() => {
+            setCtr(1);
+          }, 200);
+        }
+
+        React.useEffect(() => {
+          doSomething();
+        }, []);
+        return ctr;
+      }
+      const el = document.createElement('div');
+      await act(async () => {
+        act(() => {
+          ReactDOM.render(<App />, el);
+        });
+
+        await sleep(500);
+        expect(el.innerHTML).toBe('1');
+      });
+    });
+
+    it('warns if you do not await an act call', async () => {
+      spyOnDevAndProd(console, 'error');
+      act(async () => {});
+      // it's annoying that we have to wait a tick before this warning comes in
+      await sleep(0);
+      expect(console.error).toHaveBeenCalledTimes(1);
+    });
+
+    it('it warns if you try to interleave multiple act calls', async () => {
+      spyOnDevAndProd(console, 'error');
+      (async () => {
+        await act(async () => {
+          await sleep(200);
+        });
+      })();
+
+      await act(async () => {
+        await sleep(500);
+      });
+
+      await sleep(1000);
+      expect(console.error).toHaveBeenCalledTimes(1);
+    });
+
+    it('commits are effects are guaranteed to be flushed', async () => {
+      function App(props) {
+        let [state, setState] = React.useState(0);
+        async function something() {
+          await null;
+          setState(1);
+        }
+        React.useEffect(() => {
+          something();
+        }, []);
+        React.useEffect(() => {
+          props.callback();
+        });
+        return state;
+      }
+      let ctr = 0;
+      const div = document.createElement('div');
+
+      await act(async () => {
+        act(() => {
+          ReactDOM.render(<App callback={() => ctr++} />, div);
+        });
+        expect(div.innerHTML).toBe('0');
+        expect(ctr).toBe(1);
+      });
+
```

Comment:
this might seem weird, but it matches user behaviour exactly - a flash of '0', followed by '1'.